src/js/plugins/tts/PageChunkIterator.js

```diff
@@ -0,0 +1,153 @@
+import PageChunk from './PageChunk.js';
+
+/**
+ * Class that iterates over the page chunks of a book; caching/buffering
+ * as much as possible to try to ensure a smooth experience.
+ */
+export default class PageChunkIterator {
+    /**
+     * @param {number} pageCount total number of pages
+     * @param {number} start page to start on
+     * @param {PageChunkIteratorOptions} opts
+     */
+    constructor(pageCount, start, opts) {
+        this.pageCount = pageCount;
+        this.opts = Object.assign({}, DEFAULT_OPTS, opts);
+        /** Position in the chunk sequence */
+        this._cursor = { page: start, chunk: 0 };
+        /** @type {Object<number, PageChunk[]>} leaf index -> chunks*/
+        this._bufferedPages = {};
+        /** @type {Object<number, PromiseLike<PageChunk[]>} leaf index -> chunks*/
+        this._bufferingPages = {};
+        /**
+         * @type {Promise} promise that manages cursor modifications so that they
+         * happen in order triggered as opposed to order the server responds
+         **/
+        this._cursorLock = Promise.resolve();
+    }
+
+    /**
+     * Get the next chunk
+     * @return {PromiseLike<"__PageChunkIterator.AT_END__" | PageChunk>}
+     */
+    next() {
+        return this._cursorLock = this._cursorLock
+        .then(() => this._nextUncontrolled());
+    }
+
+    /**
+     * Sends the cursor back 1
+     * @return {Promise}
+     **/
+    decrement() {
+        return this._cursorLock = this._cursorLock
+        .then(() => this._decrementUncontrolled());
+    }
+
+    /**
+     * Gets without ensuring synchronization
+     * @return {PromiseLike<"__PageChunkIterator.AT_END__" | PageChunk>}
+     */
+    _nextUncontrolled() {
+        if (this._cursor.page == this.pageCount) {
+            return Promise.resolve(PageChunkIterator.AT_END);
+        }
+
+        this._recenterBuffer(this._cursor.page);
+
+        return this._fetchPageChunks(this._cursor.page)
+        .then(chunks => {
+            if (this._cursor.chunk == chunks.length) {
+                this._cursor.page++;
+                this._cursor.chunk = 0;
+                return this._nextUncontrolled();
+            }
+            return chunks[this._cursor.chunk++];
+        });
+    }
+
+    /**
+     * Decrements without ensuring synchronization
+     * @return {Promise}
+     */
+    _decrementUncontrolled() {
+        if (this._cursor.page == 0 && this._cursor.chunk == 0) {
```

Comment:
suggestion: use `===`
Also, can you use falsey values to make these comparator blocks work?

Reply:
I think I'd rather opt-out of this one; since these are explicitly numbers, I find it easier to read with `0`. Also I'd have to change `==` throughout a lot of places in this file if I were to follow that convention :P